### ISSUE TITLE
Unwritable object problem

### ISSUE THREAD

**Issue by shane-huang** [fmassa/optimize-net#27]
I tried this optimization on googlenet (https://github.com/soumith/imagenet-multiGPU.torch) training with SpatialBatchNormalization. I added the optnet.optimizeMemory in trainBatch (where a single mini-batch is trained), right after inputs are transferred to GPU and before SGD started. Just got below error. 

`
/home/user/shane/torch/install/bin/luajit: ...er/shane/torch/install/share/lua/5.1/threads/threads.lua:179: [thread 28 endcallback] /home/user/shane/torch/install/share/lua/5.1/torch/File.lua:141: Unwritable object <function> at <?>.<?>.modules.1.<?>.modules.1.<?>.updateGradInput.basefunc.find.get.finders.1.fallback.getConvolutionDescriptor_ffi.ffi.abi
stack traceback:
        [C]: in function 'error'
        /home/user/shane/torch/install/share/lua/5.1/torch/File.lua:141: in function 'writeObject'
...
        ...e/user/shane/torch/install/share/lua/5.1/optnet/init.lua:399: in function 'optimizeMemory'
`
It seems there's sth. not serializable.  I saw someone met similar problem but problem is with cudnn v3. I am using cudnn v5.  I already set the mode=training and inplace=true as instructions.

**Comment by fmassa** [comment]
I think your model contains a `DataParallelTable`. You should remove it before optimizing your network, and once it's optimized you add it back.

**Comment by fmassa** [comment]
@shane-huang check this example, which illustrates how to remove the `DataParallelTable`, optimizes the network, and then adds it back
https://github.com/facebook/fb.resnet.torch/blob/master/models/init.lua#L36-L47

I'm closing the issue, but feel free to ask if this doesn't work for you.

**Comment by shane-huang** [comment]
It works now. Thank you very much.

**Comment by shane-huang** [comment]
I don't know where to ask these question so I just put it here. @fmassa  Hopefully you could help answer them when you have time. :)

I found a weird problem when working with optnet. I use almost the same code as https://github.com/facebook/fb.resnet.torch/blob/master/models/init.lua#L36-L47 Just apply it to googlenet with batchNormalization.

If I call optimizeMemory before model:getParameters (position 1 in code below), the memory footprint on GPU shrunk to 80% of original; I assume this works correctly. But if I call optimizeMemory after model:getParameters(position 2 in code below), the memory footprint didn't get reduced without any explicit error. Why is this happening? 
```
//... require optnet, sampleInput initialize, etc.
// optnet.optimizeMemory(model,sampleInput,{inplace=false,mode='training'}) -- position 1
local parameters, gradParameters = model:getParameters()
 //optnet.optimizeMemory(model,sampleInput,{inplace=false,mode='training'}) -- position 2
...
```

**Comment by shane-huang** [comment]
@fmassa Another problem is,  optimizeMemory only works on the 1st epoch.  

I did optimizeMemory at the init stage of the model as in fb resnet. Multiple threads were started to load data. The training was on multiple GPUs. The GPU memory footprint of the first epoch is successfully reduced to around 80% of original as expect, but the following epochs still has full mem usage of original. No error is thrown during the whole process.  

Then I tried optimizeMemory before each epoch.  Only the 1st optimizeMemory did the work (if called before getParameters - refer to above comment); all following calls said "Skipping memory optimization. Network is already optimized for training mode". It seems the __memoryOptimized flag in model is not reset when an epoch finishes. By observing the real time status of the GPU memory usage, it seems the models are reloaded to GPUs each time a new epoch starts. It appears the storage plan changes somewhere while the flag is not cleared, which is weird. 
 
Then again I tried optimizeMemory when each epoch starts and removeOptimization after each epoch is done. At the second epoch optimizeMemory throws below error. The error seems due to the mode in BN module not set to training, which should be done at init. But I did initialize the SpatialBatchNormalization modules as you did with BNInit function. Still not sure why this is happening. 
```
/home/user/shane/torch/install/bin/luajit: .../user/shane/torch/install/share/lua/5.1/nn/Container.lua:67:
In 2 module of nn.Sequential:
In 1 module of nn.Concat:
In 5 module of nn.Sequential:
In 1 module of nn.Concat:
In 2 module of nn.Sequential:
...torch/install/share/lua/5.1/cudnn/BatchNormalization.lua:84: cudnn.BatchNormalization doesnt support backward in evaluate, use nn
stack traceback:
        [C]: in function 'assert'
        ...torch/install/share/lua/5.1/cudnn/BatchNormalization.lua:84: in function <...torch/install/share/lua/5.1/cudnn/BatchNormalization.lua:83>
...
/home/user/shane/torch/install/bin/luajit: .../user/shane/torch/install/share/lua/5.1/nn/Container.lua:67:
In 2 module of nn.Sequential:
In 1 module of nn.Concat:
In 5 module of nn.Sequential:
In 1 module of nn.Concat:
In 2 module of nn.Sequential:
...torch/install/share/lua/5.1/cudnn/BatchNormalization.lua:84: cudnn.BatchNormalization doesnt support backward in evaluate, use nn
stack traceback:
        [C]: in function 'assert'
        ...torch/install/share/lua/5.1/cudnn/BatchNormalization.lua:84: in function <...torch/install/share/lua/5.1/cudnn/BatchNormalization.lua:83>

```

**Comment by fmassa** [comment]
Just to make sure, are you using the exact sape code as `fb.resnet-torch`? In the end of each epoch, a snapshot is saved and `clearState` is called, which removes the optimizations inside the network. There is a code in the checkpointing file that handles everything properly. Let me know if you can't find it, I'm typing from my phonr

**Comment by fmassa** [comment]
Just to make sure, are you using the exact sape code as `fb.resnet-torch`? In the end of each epoch, a snapshot is saved and `clearState` is called, which resets the optimizations done inside the network. There is a code in the checkpointing file that handles everything properly. Let me know if you can't find it and I'll give you the pointer, I'm typing from my phone.

**Comment by shane-huang** [comment]
@fmassa I only use the same code as in init.lua in fb.resnet. I will look at the code in checkpoint. Maybe that will solve the issue of only 1st epoch working. But why getParameters could impact the way optimizeMemory works? If I put it after getParameters, the 1st epoch don't get reduced memory either.

**Comment by fmassa** [comment]
I don't see any reason at the moment for it not to work.
could you write a minimal isolated script that reproduces your problem on your machine, only defining the model and then calling optnet? If this works, then the problem is elsewhere in the code.

**Comment by shane-huang** [comment]
I found the reason why only 1st epoch working. The problem is with clearState. In the code I'm working with, model:clearState() is called at end of each epoch without any deepCopy backups.  So the optimizations on GPU nets are gone since the 2nd epoch.  <br>
As for getParameters() issue, I can't replicate it in your googlenet example version. I'm playing with soumith's version of googlenet training with a few customizations (https://github.com/soumith/imagenet-multiGPU.torch).  I happened to found if I put optimizeMemory before getParameters() line in train.lua(https://github.com/soumith/imagenet-multiGPU.torch/blob/master/train.lua#L141), the optimization works, otherwise it won't. Just curious about why.
